src/Utilities/MoistThermodynamics/src/MoistThermodynamics.jl

```diff
@@ -341,170 +603,333 @@ function saturation_shum(T, ρ, q_liq=0, q_ice=0)
+      q_c   = saturation_excess(T, ρ, q_tot)
+      q_liq, q_ice = phase_partitioning_eq(T, ρ, q_tot)
+      return θ_liq_ice - liquid_ice_pottemp_sat(T, p, q_tot, q_liq, q_ice)
+    end
+    T, converged = find_zero(eos_root, T_1, T_2, Tuple(1),
+                             IterParams(DT(1e-3), 10),
+                             SecantMethod())
+    return T
+  end
+end
 
-Return the liquid-ice potential temperature, given the temperature `T`,
-pressure `p`, total specific humidity `q_tot`, liquid specific humidity `q_liq`,
-and ice specific humidity `q_ice`.
 """
-function liquid_ice_pottemp(T, p, q_tot=0, q_liq=0, q_ice=0)
+    liquid_ice_pottemp(T, p[, q_tot=0, q_liq=0, q_ice=0])
 
+The liquid-ice potential temperature where
+ - `T` temperature
+ - `p` pressure
+and, optionally,
+ - `q_tot` total specific humidity
+ - `q_liq` liquid specific humidity
+ - `q_ice` ice specific humidity
+"""
+function liquid_ice_pottemp(T::DT, p::DT, q_tot::DT=DT(0), q_liq::DT=DT(0), q_ice::DT=DT(0)) where DT
     # isobaric specific heat of moist air
     _cp_m   = cp_m(q_tot, q_liq, q_ice)
 
     # liquid-ice potential temperature, approximating latent heats
     # of phase transitions as constants
-    return dry_pottemp(T, p, q_tot, q_liq, q_ice) * exp(-(LH_v0*q_liq + LH_s0*q_ice)/(_cp_m*T))
-
+    return dry_pottemp(T, p, q_tot, q_liq, q_ice) * exp(-(DT(LH_v0)*q_liq + DT(LH_s0)*q_ice)/(_cp_m*T))
 end
 
 """
-    dry_pottemp(T, p, q_tot=0, q_liq=0, q_ice=0)
+    liquid_ice_pottemp(ts::ThermodynamicState)
 
-Return the dry potential temperature, given the temperature `T`,
-pressure `p`, total specific humidity `q_tot`, liquid specific humidity `q_liq`,
-and ice specific humidity `q_ice`.
+The liquid-ice potential temperature,
+given a thermodynamic state `ts`.
 """
-dry_pottemp(T, p, q_tot=0, q_liq=0, q_ice=0) = T / exner(p, q_tot, q_liq, q_ice)
+liquid_ice_pottemp(ts::ThermodynamicState) =
+  liquid_ice_pottemp(air_temperature(ts), air_pressure(ts), ts.q_tot,
+                     phase_partitioning_eq(ts)...)
 
 """
-    exner(p, q_tot=0, q_liq=0, q_ice=0)
+    dry_pottemp(T, p, q_tot, q_liq, q_ice)
 
-Return the Exner function, given the pressure `p`, total specific
-humidity `q_tot`, liquid specific humidity `q_liq`, and ice specific humidity `q_ice`.
+The dry potential temperature where
+
+ - `T` temperature
+ - `p` pressure
+and, optionally,
+ - `q_tot` total specific humidity
+ - `q_liq` liquid specific humidity
+ - `q_ice` ice specific humidity
+ """
+dry_pottemp(T::DT, p::DT, q_tot::DT=DT(0), q_liq::DT=DT(0), q_ice::DT=DT(0)) where {DT} =
+  T / exner(p, q_tot, q_liq, q_ice)
+
+"""
+    dry_pottemp(ts::ThermodynamicState)
+
+The dry potential temperature, given a thermodynamic state `ts`.
 """
-function exner(p, q_tot=0, q_liq=0, q_ice=0)
+dry_pottemp(ts::ThermodynamicState) =
+  dry_pottemp(air_temperature(ts), air_pressure(ts), ts.q_tot,
+              phase_partitioning_eq(ts)...)
+
 
+"""
+    virtual_temp(T, p, q_tot, q_liq, q_ice)
```

Comment:
Replace `virtual_temp` by `virtual_pottemp` (because this is what is now computing)